src/garage/experiment/task_sampler.py

```diff
@@ -0,0 +1,227 @@
+"""Efficient and general interfaces for sampling tasks for Meta-RL."""
+import copy
+import math
+
+import numpy as np
+
+from garage.sampler.env_update import (ExistingEnvUpdate, NewEnvUpdate,
+                                       SetTaskUpdate)
+
+
+def _sample_indices(n_to_sample, n_available_tasks, with_replacement):
+    """Select indices of tasks to sample.
+
+    Args:
+        n_to_sample (int): Number of environments to sample. May be greater
+            than n_available_tasks.
+        n_available_tasks (int): Number of available tasks. Task indices will
+            be selected in the range [0, n_available_tasks).
+        with_replacement (bool): Whether tasks can repeat when sampled.
+            Note that if more tasks are sampled than exist, then tasks may
+            repeat, but only after every environment has been included at
+            least once in this batch. Ignored for continuous task spaces.
+
+    Returns:
+        np.ndarray[int]: Array of task indices.
+
+    """
+    if with_replacement:
+        return np.random.randint(n_available_tasks, size=n_to_sample)
+    else:
+        blocks = []
+        for _ in range(math.ceil(n_to_sample / n_available_tasks)):
+            s = np.arange(n_available_tasks)
+            np.random.shuffle(s)
+            blocks.append(s)
+        return np.concatenate(blocks)[:n_to_sample]
+
+
+class TaskSampler:
+    """Class for sampling batches of tasks, represented as `~EnvUpdate`s."""
+
+    def sample(self, n_tasks, with_replacement=False):
+        """Sample a list of environment updates.
+
+        Args:
+            n_tasks (int): Number of updates to sample.
+            with_replacement (bool): Whether tasks can repeat when sampled.
+                Note that if more tasks are sampled than exist, then tasks may
+                repeat, but only after every environment has been included at
+                least once in this batch. Ignored for continuous task spaces.
+
+        Returns:
+            list[EnvUpdate]: Batch of sampled environment updates, which, when
+                invoked on environments, will configure them with new tasks.
+                See :py:class:`~EnvUpdate` for more information.
+
+        """
+
+    @property
```

Comment:
document properties like attributes

Reply:
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

Reply:
Neat, this actually encourages me to use properties more.